proposals/fate-operator-proposal.md

```diff
@@ -0,0 +1,199 @@
+   * Kubefate: the resource management component of FATE cluster.
+2.	Example of full lifecycle to create KubeFATE component, deploy FATE cluster and submit a FML job to created FATE and get the result. Note that, KubeFATE and FATE cluster only needs to be deployed once, and can handle multiple jobs.
+
+## Non-Goals
+For the scope of this proposal, we won’t be addressing the method of serving the model.
+
+## Design
+
+### Container images and deploying FATE cluster on Kubernetes
+We have built a set of Docker images for FATE cluster, and put into: https://hub.docker.com/orgs/federatedai/repositories . All the images have been already used and verified by community users. 
+
+There is a provisioning and management component of FATE cluster, called [KubeFATE](https://github.com/FederatedAI/KubeFATE/tree/master/k8s-deploy). KubeFATE manages FATE clusters of one party in a federation.
+
+All images work well and are proven in users’ environments.
+
+### Custom Resource Definition
+#### Kubefate
+```
+apiVersion: app.kubefate.net/v1beta1
+kind: Kubefate
+metadata:
+  name: kubefate-sample
+  namespace: kube-fate
+spec:
+  # kubefate image tag
+  imageVersion: v1.0.2
+  # ingress host
+  ingressDomain: kubefate.net
+  # kubefate config
+  config:
+    FATECLOUD_MONGO_USERNAME: "root"
+    FATECLOUD_MONGO_PASSWORD: "root"
+    FATECLOUD_MONGO_DATABASE: "KubeFate"
+    FATECLOUD_REPO_NAME: "kubefate"
+    FATECLOUD_REPO_URL: "https://federatedai.github.io/KubeFATE/"
+    FATECLOUD_USER_USERNAME: "admin"
+    FATECLOUD_USER_PASSWORD: "admin"
+    FATECLOUD_LOG_LEVEL: "debug"
+```
+KubeFATE is a core component to manage and coordinate FATE clusters in one FML party. The above CRD defines the KubeFATE component. 
+* ingressDomain defines other components how to access the service ingress of KubeFATE exposed.
+
+#### FATECluster
+```
+apiVersion: app.kubefate.net/v1beta1
+kind: FateCluster
+metadata:
+  name: fatecluster-sample
+  namespace: fate-9999
+spec:
+  version: v1.4.0
+  partyId: "9999"
+  proxyPort: "30009"
+  partyList:
+  - partyId: "10000"
+    partyIp: "192.168.1.10"
+    partyPort: "30010"
+  egg:
+replica: 1
+  # Points to KubeFATE created by CRD “Kubefate”
+  kubefate:
+    name: kubefate-sample
+    namespace:  kube-fate
+```
+The FateCluster defines a deployment of FATE on Kubernetes. 
+* version defines the FATE version deployed in Kubernetes;
+* partyId defines the FML party’s ID;
+* proxyPort defines the exposed port for exchanging models and parameters between different parties in a FML training. It will be exposed as a node port;
+* partyList defines the parties in a federation which take part in collaboratively learning;
+* egg is the worker nodes of FATE.
+
+#### FateJob
+```
+apiVersion: app.kubefate.net/v1beta1
+kind: FateJob
+metadata:
+  name: fatejob-sample
+  namespace: fate-9999
+spec:
+  # fateClusterRef points to the FATE cluster resource created by FateCluster
+  fateClusterRef: fatecluster-sample
+  jobConf:
```

Comment:
This looks very different from the existing spec of various operators in Kubeflow. Could you take a look at https://github.com/kubeflow/common to see if we can align the interface?

Reply:
We planned to reuse the common APIs, but a fundamental difference is hard to solve. To FATE, we need the FATE cluster is always deployed, and ready for training requests from both self party and collative parties, which means the training pod is not on demands.